File: src/commands/blame.ts
import { execSync, spawnSync } from "node:child_process";
import { existsSync, realpathSync } from "node:fs";
import { resolve } from "node:path";
import { NOTES_REF } from "../config.ts";
import { openDb } from "../store.ts";
import { findPromptBefore } from "../transcript.ts";
import { findRepoRoot, gitBypassEnv, relPathInRepo } from "../util/git.ts";

interface BlameOptions {
  file: string;
}

interface BlameLine {
  sha: string;
  origLine: number;
  finalLine: number;
  content: string;
  // From git blame --porcelain header (only on first chunk per commit).
  author?: string;
  authorTime?: number;
  summary?: string;
}

interface MatchedRange {
  event_id: string;
  session_id: string | null;
  agent_slug: string;
  model_slug: string | null;
  event_ts: string;
  start_line: number;
  end_line: number;
}

const ZERO_SHA = "0000000000000000000000000000000000000000";

export async function runBlame(opts: BlameOptions): Promise<void> {
  const rawAbs = resolve(opts.file);
  if (!existsSync(rawAbs)) {
    process.stderr.write(`posthook blame: file not found: ${rawAbs}\n`);
    process.exit(1);
  }
  const absPath = (() => {
    try {
      return realpathSync(rawAbs);
    } catch {
      return rawAbs;
    }
  })();
  const repoRoot = findRepoRoot(absPath);
  if (!repoRoot) {
    process.stderr.write(`posthook blame: not inside a git repo: ${absPath}\n`);
    process.exit(1);
  }
  const relPath = relPathInRepo(repoRoot, absPath);
  if (!relPath) {
    process.stderr.write(`posthook blame: file is outside repo root\n`);
    process.exit(1);
  }

  let porcelain: string;
  try {
    porcelain = execSync(`git blame --porcelain ${JSON.stringify(relPath)}`, {
      cwd: repoRoot,
      encoding: "utf8",
      maxBuffer: 64 * 1024 * 1024,
      env: gitBypassEnv(),
    });
  } catch (err) {
    process.stderr.write(
      `posthook blame: git blame failed: ${err instanceof Error ? err.message : String(err)}\n`,
    );
    process.exit(1);
  }

  const lines = parsePorcelain(porcelain);
  const matches = lookupRanges(repoRoot, relPath, lines);

  printBlame(relPath, lines, matches);
}

interface NoteEntry {
  lines: string;
  agent?: string;
  session?: string | null;
  model?: string | null;
  ts?: string;
}
interface NoteBody {
  v?: number;
  commit?: string;
  files?: Record<string, NoteEntry[]>;
}

function readNoteForCommit(repoRoot: string, sha: string): NoteBody | null {
  const res = spawnSync("git", ["notes", `--ref=${NOTES_REF}`, "show", sha], {
    cwd: repoRoot,
    encoding: "utf8",
    env: gitBypassEnv(),
  });
  if (res.status !== 0 || !res.stdout) return null;
  try {
    return JSON.parse(res.stdout) as NoteBody;
  } catch {
    return null;
  }
}

function parseLineRange(spec: string): [number, number] | null {
  const m = spec.match(/^(\d+)(?:-(\d+))?$/);
  if (!m || !m[1]) return null;
  const start = parseInt(m[1], 10);
  const end = m[2] ? parseInt(m[2], 10) : start;
  return [start, end];
}

function parsePorcelain(out: string): BlameLine[] {
  const result: BlameLine[] = [];
  const rawLines = out.split("\n");
  let i = 0;
  // Track per-commit metadata; porcelain emits header fields once per commit
  // and only the SHA line for subsequent chunks of the same commit.
  const meta = new Map<
    string,
    { author?: string; authorTime?: number; summary?: string }
  >();

  while (i < rawLines.length) {
    const headerLine = rawLines[i];
    if (headerLine === undefined || headerLine === "") {
      i++;
      continue;
    }
    const header = headerLine.match(/^([0-9a-f]{40}) (\d+) (\d+)(?: (\d+))?$/);
    if (!header || !header[1] || !header[2] || !header[3]) {
      i++;
      continue;
    }
    const sha = header[1];
    const origLine = parseInt(header[2], 10);
    const finalLine = parseInt(header[3], 10);
    let author: string | undefined;
    let authorTime: number | undefined;
    let summary: string | undefined;
    i++;
    while (i < rawLines.length) {
      const cur = rawLines[i];
      if (cur === undefined || cur.startsWith("\t")) break;
      const space = cur.indexOf(" ");
      const k = space === -1 ? cur : cur.slice(0, space);
      const v = space === -1 ? "" : cur.slice(space + 1);
      if (k === "author") author = v;
      else if (k === "author-time") authorTime = parseInt(v, 10);
      else if (k === "summary") summary = v;
      i++;
    }
    const contentLine = rawLines[i];
    if (contentLine !== undefined && contentLine.startsWith("\t")) {
      const cached = meta.get(sha) ?? {};
      if (author) cached.author = author;
      if (authorTime !== undefined) cached.authorTime = authorTime;
      if (summary) cached.summary = summary;
      meta.set(sha, cached);
      result.push({
        sha,
        origLine,
        finalLine,
        content: contentLine.slice(1),
        author: cached.author,
        authorTime: cached.authorTime,
        summary: cached.summary,
      });
      i++;
    }
  }
  return result;
}

interface RangeRow {
  event_id: string;
  session_id: string | null;
  agent_slug: string;
  model_slug: string | null;
  event_ts: string;
  start_line: number;
  end_line: number;
  commit_sha: string | null;
}

function lookupRanges(
  repoRoot: string,
  relPath: string,
  lines: BlameLine[],
): Map<number, MatchedRange> {
  const db = openDb();
  const repo = db
    .query("SELECT id FROM repositories WHERE root_path = ?")
    .get(repoRoot) as { id: string } | undefined;

  let rows: RangeRow[] = [];
  if (repo) {
    rows = db
      .query(
        `WITH file_ranges AS (
           SELECT elr.event_id, elr.start_line, elr.end_line,
                  e.session_id, e.agent_slug, e.ts AS event_ts,
                  s.model_slug
           FROM event_line_ranges elr
           JOIN events e ON e.id = elr.event_id
           LEFT JOIN sessions s ON s.id = e.session_id
           WHERE e.repo_id = ? AND elr.rel_file_path = ?
         )
         SELECT fr.event_id, fr.session_id, fr.agent_slug, fr.model_slug, fr.event_ts,
                fr.start_line, fr.end_line,
                (SELECT c.sha FROM commits c
                    WHERE c.repo_id = ?
                      AND datetime(c.committed_at) >= datetime(fr.event_ts)
                    ORDER BY datetime(c.committed_at) ASC LIMIT 1) AS commit_sha
         FROM file_ranges fr
         ORDER BY datetime(fr.event_ts) ASC`,
      )
      .all(repo.id, relPath, repo.id) as RangeRow[];
  }

  // Fall back to refs/notes/posthook for any commit in this file's blame output that
  // has no local SQLite ranges. This is how blame works for teammates who clone the repo
  // and run posthook without having captured the AI events themselves.
  const localCommits = new Set(rows.map((r) => r.commit_sha).filter((s): s is string => !!s));
  const uniqueCommits = new Set(lines.map((l) => l.sha).filter((s) => s !== ZERO_SHA));
  for (const sha of uniqueCommits) {
    if (localCommits.has(sha)) continue;
    const note = readNoteForCommit(repoRoot, sha);
    if (!note?.files) continue;
    const entries = note.files[relPath];
    if (!entries) continue;
    for (const entry of entries) {
      const r = parseLineRange(entry.lines);
      if (!r) continue;
      rows.push({
        event_id: `note:${sha}:${entry.session ?? "?"}:${r[0]}-${r[1]}`,
        session_id: entry.session ?? null,
        agent_slug: entry.agent ?? "unknown",
        model_slug: entry.model ?? null,
        event_ts: entry.ts ?? "",
        start_line: r[0],
        end_line: r[1],
        commit_sha: sha,
      });
    }
  }

  if (process.env.POSTHOOK_DEBUG === "1") {
    process.stderr.write(`[posthook] blame: ${rows.length} candidate range(s)\n`);
    for (const r of rows) {
      process.stderr.write(
        `[posthook]   range lines=${r.start_line}-${r.end_line} ts=${r.event_ts} commit=${r.commit_sha?.slice(0, 7) ?? "null"}\n`,
      );
    }
  }
  // For each blame line, find the LAST AI range whose start..end covers it AND whose commit_sha matches
  // the blame's SHA. For uncommitted (zero-sha) lines, match ranges with no commit_sha yet.
  const matches = new Map<number, MatchedRange>();
  for (const line of lines) {
    const wantCommitted = line.sha !== ZERO_SHA;
    // Pick the most recent range that fits.
    let pick: RangeRow | null = null;
    for (const r of rows) {
      const inRange = r.start_line <= line.origLine && line.origLine <= r.end_line;
      if (!inRange) continue;
      if (wantCommitted) {
        if (r.commit_sha !== line.sha) continue;
      } else {
        if (r.commit_sha !== null) continue;
      }
      pick = r; // keep iterating; rows are ordered ASC so last winning row is most recent
    }
    if (pick) {
      matches.set(line.finalLine, {
        event_id: pick.event_id,
        session_id: pick.session_id,
        agent_slug: pick.agent_slug,
        model_slug: pick.model_slug,
        event_ts: pick.event_ts,
        start_line: pick.start_line,
        end_line: pick.end_line,
      });
    }
  }
  return matches;
}

function printBlame(
  relPath: string,
  lines: BlameLine[],
  matches: Map<number, MatchedRange>,
): void {
  const db = openDb();
  const prompts = resolvePromptsForEvents(db, matches);

  console.log(`posthook blame ${relPath}`);
  console.log("");
  const lineWidth = String(lines.length).length;
  const tagWidth = 28;

  let prevEventId: string | null = null;
  for (const l of lines) {
    const match = matches.get(l.finalLine);
    let tag: string;
    if (match) {
      // Show a one-line prompt header when entering a new AI-authored block.
      if (match.event_id !== prevEventId) {
        const prompt = prompts.get(match.event_id);
        if (prompt) {
          const header = formatPrompt(prompt);
          process.stdout.write(
            `  ${" ".repeat(lineWidth)}  ┌─ ${header}\n`,
          );
        }
      }
      prevEventId = match.event_id;
      const ts = match.event_ts.slice(11, 16);
      const session = match.session_id ? match.session_id.slice(0, 8) : "?";
      const model = compactModel(match.model_slug);
      tag = `AI ${model} ${ts} ${session}`;
    } else {
      prevEventId = null;
      if (l.sha === ZERO_SHA) {
        tag = "uncommitted";
      } else {
        const author = l.author ? l.author.split(" ")[0] : "?";
        tag = `human · ${author}`;
      }
    }
    const lineNum = String(l.finalLine).padStart(lineWidth);
    process.stdout.write(`  ${lineNum}  ${tag.padEnd(tagWidth)}  ${l.content}\n`);
  }
  console.log("");
  const aiLines = Array.from(matches.values()).length;
  const totalLines = lines.length;
  const pct = totalLines > 0 ? (aiLines / totalLines) * 100 : 0;
  console.log(`  ${aiLines}/${totalLines} lines AI-authored (${pct.toFixed(1)}%)`);
}

function resolvePromptsForEvents(
  db: ReturnType<typeof openDb>,
  matches: Map<number, MatchedRange>,
): Map<string, string> {
  const result = new Map<string, string>();
  const eventIds = new Set<string>();
  for (const m of matches.values()) eventIds.add(m.event_id);
  if (eventIds.size === 0) return result;

  // Bulk-load (event_id, ts, transcript_path) for the events we care about.
  const placeholders = Array.from(eventIds).map(() => "?").join(",");
  const rows = db
    .query(
      `SELECT id, ts, json_extract(payload, '$.transcript_path') AS path
       FROM events
       WHERE id IN (${placeholders})`,
    )
    .all(...eventIds) as Array<{ id: string; ts: string; path: string | null }>;

  // Cache per-transcript prompts to avoid re-reading the same JSONL multiple times.
  const cache = new Map<string, Map<string, string | null>>();
  for (const r of rows) {
    if (!r.path) continue;
    let perTranscript = cache.get(r.path);
    if (!perTranscript) {
      perTranscript = new Map();
      cache.set(r.path, perTranscript);
    }
    let prompt = perTranscript.get(r.ts);
    if (prompt === undefined) {
      prompt = findPromptBefore(r.path, r.ts);
      perTranscript.set(r.ts, prompt ?? null);
    }
    if (prompt) result.set(r.id, prompt);
  }
  return result;
}

function formatPrompt(text: string): string {
  // Strip carriage returns and collapse whitespace for one-line display.
  // Truncate to a width that fits in standard terminals.
  const cleaned = text.replace(/\s+/g, " ").trim();
  const limit = 96;
  return cleaned.length > limit ? `${cleaned.slice(0, limit - 1)}…` : cleaned;
}

function compactModel(model: string | null): string {
  if (!model) return "?";
  // Trim version suffixes for display: "claude-opus-4-7-20250101" → "claude-opus-4-7"
  const trimmed = model.replace(/-\d{8,}$/, "");
  return trimmed.length > 16 ? trimmed.slice(0, 16) : trimmed;
}

File: internal/commands/blame.go
package commands

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/bilanc/posthook/internal/gitx"
	"github.com/bilanc/posthook/internal/logx"
	"github.com/bilanc/posthook/internal/paths"
	"github.com/bilanc/posthook/internal/store"
	"github.com/bilanc/posthook/internal/transcript"

	"github.com/spf13/cobra"
)

const zeroSha = "0000000000000000000000000000000000000000"

func newBlameCmd() *cobra.Command {
	var colorMode string
	cmd := &cobra.Command{
		Use:   "blame <file>",
		Short: "Show per-line AI attribution for a file",
		Long: `Show per-line AI attribution for a file.

Every line is tagged human (from git blame), uncommitted, or AI — with the
model, time and session that wrote it, and the prompt that produced it shown
above the first line of each edit. Lines from the same agent session share a
colour. The footer lists each session with a link into the local dashboard
when it is running.`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runBlame(args[0], colorMode)
		},
	}
	cmd.Flags().StringVar(&colorMode, "color", "auto", "Colour output: auto (only on a terminal, honours NO_COLOR), always, never")
	return cmd
}

type blameLine struct {
	sha        string
	origLine   int
	finalLine  int
	content    string
	author     string
	authorTime int
	summary    string
}

type matchedRange struct {
	eventID   string
	sessionID sql.NullString
	agentSlug string
	modelSlug sql.NullString
	eventTS   string
	startLine int
	endLine   int
}

type rangeRow struct {
	eventID   string
	sessionID sql.NullString
	agentSlug string
	modelSlug sql.NullString
	eventTS   string
	startLine int
	endLine   int
	commitSHA sql.NullString
}

func runBlame(file, colorMode string) error {
	rawAbs, err := filepath.Abs(file)
	if err != nil {
		return err
	}
	if _, err := os.Stat(rawAbs); err != nil {
		return fmt.Errorf("file not found: %s", rawAbs)
	}
	absPath := rawAbs
	if real, err := filepath.EvalSymlinks(rawAbs); err == nil {
		absPath = real
	}
	repoRoot := gitx.FindRepoRoot(absPath)
	if repoRoot == "" {
		return fmt.Errorf("not inside a git repo: %s", absPath)
	}
	relPath := gitx.RelPathInRepo(repoRoot, absPath)
	if relPath == "" {
		return errors.New("file is outside repo root")
	}

	cmd := exec.Command("git", "blame", "--porcelain", relPath)
	cmd.Dir = repoRoot
	cmd.Env = gitx.BypassEnv()
	out, err := cmd.Output()
	if err != nil {
		return fmt.Errorf("git blame failed: %w", err)
	}

	lines := parsePorcelain(string(out))
	matches, err := lookupRanges(repoRoot, relPath, lines)
	if err != nil {
		return err
	}
	return printBlame(relPath, lines, matches, newPalette(colorMode))
}

var porcelainHeaderRE = regexp.MustCompile(`^([0-9a-f]{40}) (\d+) (\d+)(?: (\d+))?$`)

func parsePorcelain(out string) []blameLine {
	var result []blameLine
	rawLines := strings.Split(out, "\n")
	type commitMeta struct {
		author     string
		authorTime int
		summary    string
	}
	meta := map[string]commitMeta{}

	i := 0
	for i < len(rawLines) {
		if rawLines[i] == "" {
			i++
			continue
		}
		m := porcelainHeaderRE.FindStringSubmatch(rawLines[i])
		if m == nil {
			i++
			continue
		}
		sha := m[1]
		origLine, _ := strconv.Atoi(m[2])
		finalLine, _ := strconv.Atoi(m[3])
		var author, summary string
		var authorTime int
		i++
		for i < len(rawLines) && !strings.HasPrefix(rawLines[i], "\t") {
			cur := rawLines[i]
			sp := strings.IndexByte(cur, ' ')
			k := cur
			v := ""
			if sp != -1 {
				k = cur[:sp]
				v = cur[sp+1:]
			}
			switch k {
			case "author":
				author = v
			case "author-time":
				authorTime, _ = strconv.Atoi(v)
			case "summary":
				summary = v
			}
			i++
		}
		if i < len(rawLines) && strings.HasPrefix(rawLines[i], "\t") {
			cached := meta[sha]
			if author != "" {
				cached.author = author
			}
			if authorTime != 0 {
				cached.authorTime = authorTime
			}
			if summary != "" {
				cached.summary = summary
			}
			meta[sha] = cached
			result = append(result, blameLine{
				sha:        sha,
				origLine:   origLine,
				finalLine:  finalLine,
				content:    rawLines[i][1:],
				author:     cached.author,
				authorTime: cached.authorTime,
				summary:    cached.summary,
			})
			i++
		}
	}
	return result
}

type noteEntry struct {
	Lines   string `json:"lines"`
	Agent   string `json:"agent,omitempty"`
	Session string `json:"session,omitempty"`
	Model   string `json:"model,omitempty"`
	TS      string `json:"ts,omitempty"`
}
type noteBody struct {
	V      int                    `json:"v"`
	Commit string                 `json:"commit"`
	Files  map[string][]noteEntry `json:"files"`
}

func readNoteForCommit(repoRoot, sha string) *noteBody {
	cmd := exec.Command("git", "notes", "--ref="+paths.NotesRef, "show", sha)
	cmd.Dir = repoRoot
	cmd.Env = gitx.BypassEnv()
	var out bytes.Buffer
	cmd.Stdout = &out
	if err := cmd.Run(); err != nil {
		return nil
	}
	var n noteBody
	if err := json.Unmarshal(out.Bytes(), &n); err != nil {
		return nil
	}
	return &n
}

var lineRangeRE = regexp.MustCompile(`^(\d+)(?:-(\d+))?$`)

func parseLineRange(spec string) (int, int, bool) {
	m := lineRangeRE.FindStringSubmatch(spec)
	if m == nil {
		return 0, 0, false
	}
	start, _ := strconv.Atoi(m[1])
	end := start
	if m[2] != "" {
		end, _ = strconv.Atoi(m[2])
	}
	return start, end, true
}

func lookupRanges(repoRoot, relPath string, lines []blameLine) (map[int]matchedRange, error) {
	db, err := store.Open()
	if err != nil {
		return nil, err
	}

	var repoID sql.NullString
	_ = db.QueryRow(`SELECT id FROM repositories WHERE root_path = ?`, repoRoot).Scan(&repoID)

	var rows []rangeRow
	if repoID.Valid {
		queryRows, err := db.Query(`
			WITH file_ranges AS (
				SELECT elr.event_id, elr.start_line, elr.end_line,
				       e.session_id, e.agent_slug, e.ts AS event_ts,
				       s.model_slug
				FROM event_line_ranges elr
				JOIN events e ON e.id = elr.event_id
				LEFT JOIN sessions s ON s.id = e.session_id
				WHERE e.repo_id = ? AND elr.rel_file_path = ?
			)
			SELECT fr.event_id, fr.session_id, fr.agent_slug, fr.model_slug, fr.event_ts,
			       fr.start_line, fr.end_line,
			       (SELECT c.sha FROM commits c
				WHERE c.repo_id = ?
				  AND datetime(c.committed_at) >= datetime(fr.event_ts)
				ORDER BY datetime(c.committed_at) ASC LIMIT 1) AS commit_sha
			FROM file_ranges fr
			ORDER BY datetime(fr.event_ts) ASC`,
			repoID.String, relPath, repoID.String)
		if err != nil {
			return nil, err
		}
		for queryRows.Next() {
			var r rangeRow
			if err := queryRows.Scan(&r.eventID, &r.sessionID, &r.agentSlug, &r.modelSlug,
				&r.eventTS, &r.startLine, &r.endLine, &r.commitSHA); err != nil {
				queryRows.Close()
				return nil, err
			}
			rows = append(rows, r)
		}
		queryRows.Close()
	}

	// Note-fallback for commits without local ranges (e.g. teammate ran
	// blame after cloning).
	local := map[string]bool{}
	for _, r := range rows {
		if r.commitSHA.Valid {
			local[r.commitSHA.String] = true
		}
	}
	uniqueCommits := map[string]bool{}
	for _, l := range lines {
		if l.sha != zeroSha {
			uniqueCommits[l.sha] = true
		}
	}
	for sha := range uniqueCommits {
		if local[sha] {
			continue
		}
		note := readNoteForCommit(repoRoot, sha)
		if note == nil || note.Files == nil {
			continue
		}
		entries := note.Files[relPath]
		for _, entry := range entries {
			s, e, ok := parseLineRange(entry.Lines)
			if !ok {
				continue
			}
			r := rangeRow{
				eventID:   fmt.Sprintf("note:%s:%s:%d-%d", sha, entry.Session, s, e),
				agentSlug: entry.Agent,
				eventTS:   entry.TS,
				startLine: s,
				endLine:   e,
				commitSHA: sql.NullString{String: sha, Valid: true},
			}
			if entry.Session != "" {
				r.sessionID = sql.NullString{String: entry.Session, Valid: true}
			}
			if entry.Model != "" {
				r.modelSlug = sql.NullString{String: entry.Model, Valid: true}
			}
			if r.agentSlug == "" {
				r.agentSlug = "unknown"
			}
			rows = append(rows, r)
		}
	}

	if os.Getenv("POSTHOOK_DEBUG") == "1" {
		logx.Debugf("blame: %d candidate range(s)", len(rows))
		for _, r := range rows {
			shaShort := "null"
			if r.commitSHA.Valid && len(r.commitSHA.String) >= 7 {
				shaShort = r.commitSHA.String[:7]
			}
			logx.Debugf("  range lines=%d-%d ts=%s commit=%s",
				r.startLine, r.endLine, r.eventTS, shaShort)
		}
	}

	// For each blame line, find the most recent range that fits AND whose
	// commit_sha matches the blame's SHA. Uncommitted (zero-sha) lines match
	// ranges with no commit_sha yet.
	matches := map[int]matchedRange{}
	for _, line := range lines {
		wantCommitted := line.sha != zeroSha
		var pick *rangeRow
		for i := range rows {
			r := &rows[i]
			if r.startLine > line.origLine || r.endLine < line.origLine {
				continue
			}
			if wantCommitted {
				if !r.commitSHA.Valid || r.commitSHA.String != line.sha {
					continue
				}
			} else {
				if r.commitSHA.Valid {
					continue
				}
			}
			pick = r // last winning row in ASC order = most recent
		}
		if pick != nil {
			matches[line.finalLine] = matchedRange{
				eventID:   pick.eventID,
				sessionID: pick.sessionID,
				agentSlug: pick.agentSlug,
				modelSlug: pick.modelSlug,
				eventTS:   pick.eventTS,
				startLine: pick.startLine,
				endLine:   pick.endLine,
			}
		}
	}
	return matches, nil
}

// palette holds the ANSI sequences used by blame; every field is empty when
// colour is off so the printing code needs no branches.
type palette struct {
	on       bool
	reset    string
	dim      string
	bold     string
	human    string
	pending  string
	sessions []string
}

func newPalette(mode string) palette {
	on := false
	switch mode {
	case "always":
		on = true
	case "never":
		on = false
	default:
		on = stdoutIsTerminal() && os.Getenv("NO_COLOR") == "" && os.Getenv("TERM") != "dumb"
	}
	if !on {
		return palette{}
	}
	return palette{
		on:      true,
		reset:   "\x1b[0m",
		dim:     "\x1b[2m",
		bold:    "\x1b[1m",
		human:   "\x1b[2m",
		pending: "\x1b[33m",
		// One colour per AI session, in order of first appearance, so edits
		// from the same session read as a block.
		sessions: []string{"\x1b[36m", "\x1b[35m", "\x1b[32m", "\x1b[34m", "\x1b[96m", "\x1b[95m", "\x1b[92m", "\x1b[94m"},
	}
}

func stdoutIsTerminal() bool {
	fi, err := os.Stdout.Stat()
	if err != nil {
		return false
	}
	return fi.Mode()&os.ModeCharDevice != 0
}

func (p palette) paint(color, text string) string {
	if !p.on || color == "" {
		return text
	}
	return color + text + p.reset
}

// sessionSummary is one row of the footer: an AI session that owns at least
// one line of the file.
type sessionSummary struct {
	key      string // session id, or agent slug when the session is unknown
	id       sql.NullString
	agent    string
	model    sql.NullString
	lines    int
	firstSeq int
	color    string
}

func printBlame(relPath string, lines []blameLine, matches map[int]matchedRange, pal palette) error {
	db, err := store.Open()
	if err != nil {
		return err
	}
	prompts := resolvePromptsForEvents(db, matches)

	// Assign each session a stable colour in order of first appearance.
	sessions := map[string]*sessionSummary{}
	var order []*sessionSummary
	for _, l := range lines {
		m, ok := matches[l.finalLine]
		if !ok {
			continue
		}
		key := m.agentSlug
		if m.sessionID.Valid {
			key = m.sessionID.String
		}
		ss, seen := sessions[key]
		if !seen {
			ss = &sessionSummary{key: key, id: m.sessionID, agent: m.agentSlug, model: m.modelSlug, firstSeq: len(order)}
			if len(pal.sessions) > 0 {
				ss.color = pal.sessions[len(order)%len(pal.sessions)]
			}
			sessions[key] = ss
			order = append(order, ss)
		}
		ss.lines++
		if !ss.model.Valid && m.modelSlug.Valid {
			ss.model = m.modelSlug
		}
	}

	fmt.Printf("%s %s\n\n", pal.paint(pal.bold, "posthook blame"), relPath)
	lineWidth := len(strconv.Itoa(len(lines)))
	tagWidth := 28
	prevEventID := ""

	for _, l := range lines {
		match, hasMatch := matches[l.finalLine]
		num := pal.paint(pal.dim, fmt.Sprintf("%*d", lineWidth, l.finalLine))
		var tag string
		if hasMatch {
			key := match.agentSlug
			if match.sessionID.Valid {
				key = match.sessionID.String
			}
			color := sessions[key].color
			if match.eventID != prevEventID {
				if prompt := prompts[match.eventID]; prompt != "" {
					fmt.Printf("  %s  %s\n", strings.Repeat(" ", lineWidth),
						pal.paint(color+pal.bold, "┌─ "+formatPrompt(prompt)))
				}
			}
			prevEventID = match.eventID
			ts := match.eventTS
			if len(ts) >= 16 {
				ts = ts[11:16]
			}
			session := "?"
			if match.sessionID.Valid && len(match.sessionID.String) >= 8 {
				session = match.sessionID.String[:8]
			} else if match.sessionID.Valid {
				session = match.sessionID.String
			}
			plain := fmt.Sprintf("AI %s %s %s", compactModel(match.modelSlug), ts, session)
			tag = pal.paint(color, fmt.Sprintf("%-*s", tagWidth, plain))
		} else {
			prevEventID = ""
			var plain, color string
			if l.sha == zeroSha {
				plain, color = "uncommitted", pal.pending
			} else {
				author := "?"
				if l.author != "" {
					author = strings.SplitN(l.author, " ", 2)[0]
				}
				plain, color = "human · "+author, pal.human
			}
			tag = pal.paint(color, fmt.Sprintf("%-*s", tagWidth, plain))
		}
		fmt.Printf("  %s  %s  %s\n", num, tag, l.content)
	}
	fmt.Println()
	aiCount := len(matches)
	total := len(lines)
	pct := 0.0
	if total > 0 {
		pct = float64(aiCount) / float64(total) * 100
	}
	fmt.Printf("  %s\n", pal.paint(pal.bold, fmt.Sprintf("%d/%d lines AI-authored (%.1f%%)", aiCount, total, pct)))

	if len(order) > 0 {
		printSessionFooter(db, order, pal)
	}
	return nil
}

// printSessionFooter lists every AI session that owns lines in the file,
// with a link into the local dashboard for sessions this machine has
// captured (a teammate's session, known only from the git note, gets no link).
func printSessionFooter(db *store.DB, order []*sessionSummary, pal palette) {
	dash := resolveDashConfig()
	dashUp := portOpen(dash.addr())

	fmt.Println()
	for _, ss := range order {
		id := "?"
		if ss.id.Valid && len(ss.id.String) >= 8 {
			id = ss.id.String[:8]
		} else if ss.id.Valid {
			id = ss.id.String
		}
		who := ss.agent
		if ss.model.Valid && ss.model.String != "" {
			who += " · " + compactModel(ss.model)
		}
		unit := "lines"
		if ss.lines == 1 {
			unit = "line"
		}
		row := fmt.Sprintf("  %s %s  %-34s %4d %s", pal.paint(ss.color, "■"), pal.paint(ss.color+pal.bold, id), who, ss.lines, unit)
		if dashUp && ss.id.Valid && sessionKnownLocally(db, ss.id.String) {
			row += "  " + pal.paint(pal.dim, dash.url()+"/sessions/"+ss.id.String)
		}
		fmt.Println(row)
	}
}

func sessionKnownLocally(db *store.DB, sessionID string) bool {
	var one int
	return db.QueryRow(`SELECT 1 FROM sessions WHERE id = ?`, sessionID).Scan(&one) == nil
}

// resolvePromptsForEvents finds, for each matched event, the prompt the user
// typed before it: first from the Claude Code transcript on disk, then from
// the prompts posthook stored at session end (transcripts are pruned by the
// agents after a while; the stored copy is not).
func resolvePromptsForEvents(db *store.DB, matches map[int]matchedRange) map[string]string {
	out := map[string]string{}
	if len(matches) == 0 {
		return out
	}
	byID := map[string]matchedRange{}
	for _, m := range matches {
		byID[m.eventID] = m
	}

	placeholders := make([]string, 0, len(byID))
	params := make([]any, 0, len(byID))
	for id := range byID {
		placeholders = append(placeholders, "?")
		params = append(params, id)
	}
	q := fmt.Sprintf(`
		SELECT id, ts, json_extract(payload, '$.transcript_path') AS path
		FROM events
		WHERE id IN (%s)`, strings.Join(placeholders, ","))
	rows, err := db.Query(q, params...)
	if err == nil {
		cache := map[string]map[string]string{}
		for rows.Next() {
			var id, ts string
			var path sql.NullString
			if err := rows.Scan(&id, &ts, &path); err != nil {
				break
			}
			if !path.Valid {
				continue
			}
			perTranscript, ok := cache[path.String]
			if !ok {
				perTranscript = map[string]string{}
				cache[path.String] = perTranscript
			}
			prompt, found := perTranscript[ts]
			if !found {
				prompt = transcript.FindPromptBefore(path.String, ts)
				perTranscript[ts] = prompt
			}
			if prompt != "" {
				out[id] = prompt
			}
		}
		rows.Close()
	}

	for id, m := range byID {
		if out[id] != "" || !m.sessionID.Valid || m.eventTS == "" {
			continue
		}
		var text string
		err := db.QueryRow(`
			SELECT prompt_text FROM session_prompts
			WHERE session_id = ? AND ts IS NOT NULL AND ts < ?
			ORDER BY ts DESC LIMIT 1`, m.sessionID.String, m.eventTS).Scan(&text)
		if err == nil && strings.TrimSpace(text) != "" {
			out[id] = text
		} else if err != nil {
			logx.Debugf("blame: stored-prompt lookup for session %s before %s: %v", m.sessionID.String, m.eventTS, err)
		}
	}
	return out
}

func formatPrompt(text string) string {
	cleaned := strings.Join(strings.Fields(text), " ")
	const limit = 96
	if len(cleaned) > limit {
		return cleaned[:limit-1] + "…"
	}
	return cleaned
}

var versionSuffixRE = regexp.MustCompile(`-\d{8,}$`)

func compactModel(m sql.NullString) string {
	if !m.Valid || m.String == "" {
		return "?"
	}
	trimmed := versionSuffixRE.ReplaceAllString(m.String, "")
	if len(trimmed) > 16 {
		return trimmed[:16]
	}
	return trimmed
}
